File: examples/Classes.java
// ------------
// Classes.java
// ------------

class MyComplex {
    private int _r;
    private int _i;

    public MyComplex () {
        _r = 0;
        _i = 0;}

    public MyComplex (int real) {
        _r = real;
        _i = 0;}

    public MyComplex (int real, int imag) {
        _r = real;
        _i = imag;}

    public MyComplex conjugate () {
        _i = -_i;
        return this;}

    public boolean equals (Object rhs) {
        if (!(rhs instanceof MyComplex))
            return false;
        MyComplex that = (MyComplex) rhs;
        return (_r == that._r) && (_i == that._i);}

    public String toString () {
        return "(" + _r + ", " + _i + ")";}}

final class Classes {
    public static void main (String[] args) {
        System.out.println("Classes.java");

        final MyComplex x = new MyComplex();
        final MyComplex y = new MyComplex(2);

        final MyComplex z = new MyComplex(2, 3);
        assert z.equals(new MyComplex(2, 3));
        assert z.toString().equals("(2, 3)");

        final MyComplex t = z.conjugate();
        assert z.equals(new MyComplex(2, -3));
        assert t == z;

        System.out.println("Done.");}}

File: examples/Classes.py
#!/usr/bin/env python

# ----------
# Classes.py
# ----------

print "Classes.py"

class My_Complex (object) :
    def __init__ (self, r = 0, i = 0) :
        self.r = r
        self.i = i

    def __eq__ (self, rhs) :
        if not isinstance(rhs, My_Complex) :
            return False
        return (self.r == rhs.r) and (self.i == rhs.i)

    def __str__ (self) :
        return "(" + str(self.r) + ", " + str(self.i) + ")"

    def conjugate (self) :
        self.i = -self.i
        return self

def conjugate (v) :
    return My_Complex(v.r, v.i).conjugate()

x = My_Complex()
y = My_Complex(2)

z = My_Complex(2, 3)
assert z is not My_Complex(2, 3)
assert z ==     My_Complex(2, 3)
assert str(z) == "(2, 3)"

t = z.conjugate()
assert z == My_Complex(2, -3)
assert z is t

t = My_Complex.conjugate(z)
assert z == My_Complex(2, 3)
assert z is t

t = conjugate(z)
assert z == My_Complex(2,  3)
assert t == My_Complex(2, -3)

print "Done."
